File: backend/app/services/websocket.py
"""
WebSocket event handlers for real-time communication
"""
from flask_socketio import emit, join_room, leave_room
from app import socketio
import random
from datetime import datetime

@socketio.on('connect')
def handle_connect():
    """Handle client connection"""
    print('Client connected')
    emit('connection_response', {'status': 'connected', 'message': 'Connected to APRU40 server'})

@socketio.on('disconnect')
def handle_disconnect():
    """Handle client disconnection"""
    print('Client disconnected')

@socketio.on('join')
def handle_join(data):
    """Join a specific room for targeted updates"""
    room = data.get('room')
    if room:
        join_room(room)
        emit('joined', {'room': room})

@socketio.on('leave')
def handle_leave(data):
    """Leave a room"""
    room = data.get('room')
    if room:
        leave_room(room)
        emit('left', {'room': room})

@socketio.on('request_update')
def handle_update_request(data):
    """Handle manual update request"""
    # Send mock real-time data
    emit('device:heartbeat', {
        'device_id': 'GW-01-N01',
        'status': 'online',
        'timestamp': datetime.utcnow().isoformat()
    })

# Helper functions to emit events (called from other parts of the app)
def emit_device_status_change(device_id, status):
    """Emit device status change event"""
    socketio.emit('device:status_change', {
        'device_id': device_id,
        'status': status,
        'timestamp': datetime.utcnow().isoformat()
    })

def emit_new_alert(alert):
    """Emit new alert event"""
    socketio.emit('alert:new', {
        'id': alert.id,
        'device_id': alert.device_id,
        'alert_type': alert.alert_type,
        'severity': alert.severity,
        'message': alert.message,
        'timestamp': alert.created_at.isoformat()
    })

def emit_sensor_data(device_id, sensor_data):
    """Emit sensor data update"""
    socketio.emit('device:sensor_data', {
        'device_id': device_id,
        'data': sensor_data,
        'timestamp': datetime.utcnow().isoformat()
    })

File: frontend/src/services/websocket.ts
/**
 * WebSocket service for real-time communication
 */
import { io, Socket } from 'socket.io-client';

const SOCKET_URL = process.env.REACT_APP_API_URL || 'http://localhost:5000';

class WebSocketService {
  private socket: Socket | null = null;
  private listeners: Map<string, Function[]> = new Map();

  connect() {
    if (this.socket?.connected) {
      return;
    }

    this.socket = io(SOCKET_URL, {
      transports: ['websocket', 'polling'],
    });

    this.socket.on('connect', () => {
      console.log('WebSocket connected');
    });

    this.socket.on('disconnect', () => {
      console.log('WebSocket disconnected');
    });

    this.socket.on('connection_response', (data) => {
      console.log('Connection response:', data);
    });

    // Set up event listeners
    this.setupEventListeners();
  }

  disconnect() {
    if (this.socket) {
      this.socket.disconnect();
      this.socket = null;
    }
  }

  private setupEventListeners() {
    const events = [
      'device:heartbeat',
      'device:status_change',
      'device:sensor_data',
      'alert:new',
      'device:config_applied',
    ];

    events.forEach((event) => {
      this.socket?.on(event, (data) => {
        console.log(`WebSocket event: ${event}`, data);
        this.notifyListeners(event, data);
      });
    });
  }

  on(event: string, callback: Function) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, []);
    }
    this.listeners.get(event)?.push(callback);
  }

  off(event: string, callback: Function) {
    const eventListeners = this.listeners.get(event);
    if (eventListeners) {
      const index = eventListeners.indexOf(callback);
      if (index > -1) {
        eventListeners.splice(index, 1);
      }
    }
  }

  private notifyListeners(event: string, data: any) {
    const eventListeners = this.listeners.get(event);
    if (eventListeners) {
      eventListeners.forEach((callback) => callback(data));
    }
  }

  joinRoom(room: string) {
    this.socket?.emit('join', { room });
  }

  leaveRoom(room: string) {
    this.socket?.emit('leave', { room });
  }

  requestUpdate() {
    this.socket?.emit('request_update', {});
  }
}

export const websocketService = new WebSocketService();
export default websocketService;
